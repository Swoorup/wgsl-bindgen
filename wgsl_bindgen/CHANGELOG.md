# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.13.2](https://github.com/Swoorup/wgsl-bindgen/compare/wgsl_bindgen-v0.13.1...wgsl_bindgen-v0.13.2) - 2024-07-05

### Other
- Bind Group Entry Collections can be cloned

## [0.13.1](https://github.com/Swoorup/wgsl-bindgen/compare/wgsl_bindgen-v0.13.0...wgsl_bindgen-v0.13.1) - 2024-07-05

### Other
- Fix `min_binding_size` when invoking entry module contains symbols.

## [0.13.0](https://github.com/Swoorup/wgsl-bindgen/compare/wgsl_bindgen-v0.12.0...wgsl_bindgen-v0.13.0) - 2024-07-05

### Other
- Added release-plz workflow
- Reference version from workspace root
- For bindgroup generation, rename Layout to EntryCollection with helper structs
- Added `min_binding_size` for buffer types where possible
- Allow to fully qualify or use relative name
- Added a way to skip generating `_root_`

## v0.12.0 (2024-06-10)

<csr-id-e52a9dbe660a417afa371f480be161d58f1dd642/>

### Other

 - <csr-id-e52a9dbe660a417afa371f480be161d58f1dd642/> format builder error message into bindgen error

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 51 commits contributed to the release over the course of 125 calendar days.
 - 1 commit was understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Added changelog ([`cd55d10`](https://github.com/Swoorup/wgsl-bindgen/commit/cd55d10c57f1e159a0c31988c67559b559a68ace))
    - Release wgsl_bindgen v0.12.0 ([`d61fd9e`](https://github.com/Swoorup/wgsl-bindgen/commit/d61fd9e174877500ba86d089101ecba7c1b5886f))
    - Fix typo ([`22adeec`](https://github.com/Swoorup/wgsl-bindgen/commit/22adeece762ad8835a812fc448a3281ae6ce42f9))
    - Added non-working support for overridable constants ([`e1937d6`](https://github.com/Swoorup/wgsl-bindgen/commit/e1937d661f920812e3587d2cb70362cad15a613f))
    - Initial upgrade to wgpu 0.20 ([`92bf827`](https://github.com/Swoorup/wgsl-bindgen/commit/92bf8274c3bdc39e4332f558a653647be61c3d95))
    - Make the texture sample type filterable ([`0660ee1`](https://github.com/Swoorup/wgsl-bindgen/commit/0660ee19a21e65f6da14835fd9cd85924ae762b1))
    - Consolidate specifying versions in the root manifest ([`42d2822`](https://github.com/Swoorup/wgsl-bindgen/commit/42d2822da5a85e1964b4442db090a6991a5b30c3))
    - Added option to change the visibily of the export types ([`88fd877`](https://github.com/Swoorup/wgsl-bindgen/commit/88fd877fc2c75c35dee3d313d93d93e22ffcb75b))
    - Fix issues with texture_2d of type i32 or u32 ([`53c0c63`](https://github.com/Swoorup/wgsl-bindgen/commit/53c0c63f6e4ea2a2569182bea2e99874ca64461e))
    - Use the renamed crate include_absolute_path ([`6f485bf`](https://github.com/Swoorup/wgsl-bindgen/commit/6f485bf0beb05992d8d2a2ee1950738fd2e434fe))
    - Make SHADER_STRING public ([`ce4f68b`](https://github.com/Swoorup/wgsl-bindgen/commit/ce4f68b418241c3224240bab42e9cbe0bae52905))
    - Regex for all overrides ([`8ea7ffd`](https://github.com/Swoorup/wgsl-bindgen/commit/8ea7ffd65871af95aaeaff8da9d4589f20ff049c))
    - Simplify also for bulk options ([`d45d6f0`](https://github.com/Swoorup/wgsl-bindgen/commit/d45d6f0898c52fa7f8ad41abb7f466e6ae2aec25))
    - Adding custom padding field support ([`998f7a8`](https://github.com/Swoorup/wgsl-bindgen/commit/998f7a8f60b83424fff93e471f04adf7130a8f83))
    - Adjust size if custom alignment is specified. ([`a4b61c7`](https://github.com/Swoorup/wgsl-bindgen/commit/a4b61c7d52496499b92b029a3604053d2420b147))
    - Ability to override alignment for structs ([`cd26b91`](https://github.com/Swoorup/wgsl-bindgen/commit/cd26b91be29870ac629a1674a8a43ba98d46b6d6))
    - Use Result type for create_shader* when using `UseComposerWithPath` ([`80a7f95`](https://github.com/Swoorup/wgsl-bindgen/commit/80a7f9594330b6e982bb91bb12991df8b79cba70))
    - Seperate types, assertions, impls in generated output ([`c2c4dc9`](https://github.com/Swoorup/wgsl-bindgen/commit/c2c4dc956925aedef11d706cd7024c8b25593a66))
    - RustSourceItem => RustItem ([`ce2a91e`](https://github.com/Swoorup/wgsl-bindgen/commit/ce2a91eca61507ba237fd9828a84a5d00a6e2d99))
    - Pass entry point name to builders ([`4fc895b`](https://github.com/Swoorup/wgsl-bindgen/commit/4fc895bef6ce8a29b32611fc363ea68a40b60405))
    - Export quote, syn functions and macros ([`782f481`](https://github.com/Swoorup/wgsl-bindgen/commit/782f481c70bb5d8ae8381c0ddf83ec4ddc6a2a79))
    - Added extra bindings generator as prep for targetting non-wgpu libs ([`9b6204d`](https://github.com/Swoorup/wgsl-bindgen/commit/9b6204d62b4daa5f45c7d9a0ee05d41380f37650))
    - Added custom field mappings ([`4132659`](https://github.com/Swoorup/wgsl-bindgen/commit/4132659692ea4a34a7cf510829a470dc3390b269))
    - Avoid HashMap for more consitent shader bindings generation ([`fd6d144`](https://github.com/Swoorup/wgsl-bindgen/commit/fd6d144dafbcc6e234d479f5c7e5c53c93f0816c))
    - Rename ShaderRegistry to ShaderEntry in output ([`1461393`](https://github.com/Swoorup/wgsl-bindgen/commit/1461393b0710e23a028478f1df131191f2398c2e))
    - Added mandatory workspace root option used for resolving imports ([`d20d3d5`](https://github.com/Swoorup/wgsl-bindgen/commit/d20d3d5176984f305d4a3e190500c4601671af85))
    - Add shader labels ([`c8a129b`](https://github.com/Swoorup/wgsl-bindgen/commit/c8a129bc5529a468eb29687b20ce4c40e6fa647f))
    - Feature shader registry and shader defines ([`187c7f4`](https://github.com/Swoorup/wgsl-bindgen/commit/187c7f417ec9be4543168c462ed6d171ba3180c6))
    - Added multiple shader source option ([`db90739`](https://github.com/Swoorup/wgsl-bindgen/commit/db90739cec926b464eb6fafb8f1254c42ad91201))
    - Add ability to override struct and path based source type ([`1d4ee0a`](https://github.com/Swoorup/wgsl-bindgen/commit/1d4ee0a552ffe4e6a9298f183bd3c9b617635908))
    - Short const constructors and fix demangle in comments ([`a49be89`](https://github.com/Swoorup/wgsl-bindgen/commit/a49be89ca98ca65ca296717b0f98e24530ad11b0))
    - Rename Capabilities to WgslShaderIRCapabilities, and update test ([`1cad0cb`](https://github.com/Swoorup/wgsl-bindgen/commit/1cad0cbe5ff581810b770c6fb95940f1472c7fd1))
    - Reexport Capabilities ([`7262606`](https://github.com/Swoorup/wgsl-bindgen/commit/7262606a6d0880c9f8aa8872197a3e151a16975b))
    - Allow setting capabilities ([`b6df117`](https://github.com/Swoorup/wgsl-bindgen/commit/b6df117b40909cfeb803c6a7782ab2d2dc906176))
    - Release new version ([`ec3d554`](https://github.com/Swoorup/wgsl-bindgen/commit/ec3d55412002d27c48200261b8e9853e9bfe8af2))
    - Make naga oil's error more useful ([`6a1bc45`](https://github.com/Swoorup/wgsl-bindgen/commit/6a1bc45524ffeb4386ff18f846588cf6c1ea0e1b))
    - Format builder error message into bindgen error ([`e52a9db`](https://github.com/Swoorup/wgsl-bindgen/commit/e52a9dbe660a417afa371f480be161d58f1dd642))
    - Ignore snake case warnings if struct is not camel case ([`54c563e`](https://github.com/Swoorup/wgsl-bindgen/commit/54c563eb3d89d9815d7391b599c1a86de3a14d25))
    - Minor corrections ([`194b3e4`](https://github.com/Swoorup/wgsl-bindgen/commit/194b3e4a66bfaad0ebc577670b50eec372701e35))
    - Added a mechanism to scan additional source directory ([`300a3d7`](https://github.com/Swoorup/wgsl-bindgen/commit/300a3d7aec20556712bd835d71a42ca375ae1da9))
    - Allow to use naga_oil compose in the generated output ([`f32c279`](https://github.com/Swoorup/wgsl-bindgen/commit/f32c279c02ea7760ce901533013f6d0da51674c5))
    - Fix direct item wgsl imports. ([`3e58108`](https://github.com/Swoorup/wgsl-bindgen/commit/3e581089e21b245bd85feecdc94f3f1d9310aacc))
    - Added failing test for direct path import for nested type ([`e014d4b`](https://github.com/Swoorup/wgsl-bindgen/commit/e014d4b6c5326a40d59291be96e24a3fd150d746))
    - Demangle bindgroup struct fields if imported from other wgsl files ([`7231f78`](https://github.com/Swoorup/wgsl-bindgen/commit/7231f78806e75a18af9f78005c3b016f16dcf1dc))
    - Add support for scalar types in bindings ([`4af047a`](https://github.com/Swoorup/wgsl-bindgen/commit/4af047aa976252211f31f882db8b5006fecb1977))
    - Add support for path based import. ([`d1e861d`](https://github.com/Swoorup/wgsl-bindgen/commit/d1e861dacd5cb04f1b74065448fde980cfc696b6))
    - Demangle name for consts items ([`5ec2c1a`](https://github.com/Swoorup/wgsl-bindgen/commit/5ec2c1a22c2b4c1855dee3d2d88fa0b46ad88d6c))
    - Updated docs, use stable features only ([`06401c5`](https://github.com/Swoorup/wgsl-bindgen/commit/06401c5eb0c5d867bee4aedf4b339f9cd373f9a5))
    - Support naga oil flavour of wgsl ([`99ea17c`](https://github.com/Swoorup/wgsl-bindgen/commit/99ea17c17bf682dd1ed9990341fb1a3aa119a6f6))
    - Enable Runtime Sized Array, Padding for bytemuck mode ([`9e21d1d`](https://github.com/Swoorup/wgsl-bindgen/commit/9e21d1dbe084f1588d7e03e2c93642ca3ffb2c05))
    - Create a fork ([`1c99e10`](https://github.com/Swoorup/wgsl-bindgen/commit/1c99e103625154dde0e357419f064e941e156f54))
</details>

