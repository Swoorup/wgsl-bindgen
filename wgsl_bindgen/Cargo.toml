[package]
name = "wgsl_bindgen"
authors.workspace = true
version.workspace = true
edition.workspace = true
repository.workspace = true
documentation.workspace = true
license.workspace = true
description = "Type safe Rust bindings workflow for wgsl shaders in wgpu"
readme = "../README.md"

[dependencies]
naga = { workspace = true, features = ["wgsl-in"] }
wgpu-types.workspace = true
syn.workspace = true
quote.workspace = true
proc-macro2.workspace = true
prettyplease.workspace = true
thiserror.workspace = true
case.workspace = true
naga_oil.workspace = true
regex.workspace = true
data-encoding.workspace = true
indexmap.workspace = true
smallvec.workspace = true
derive_more.workspace = true
fxhash.workspace = true
derivative.workspace = true
smol_str.workspace = true
colored.workspace = true
derive_builder.workspace = true
miette.workspace = true
blake3.workspace = true
regex-syntax.workspace = true
strum.workspace = true
strum_macros.workspace = true
pathdiff.workspace = true
enumflags2.workspace = true
heck.workspace = true

[dev-dependencies]
indoc.workspace = true
pretty_assertions.workspace = true
miette = { workspace = true, features = ["fancy", "syntect"] }
