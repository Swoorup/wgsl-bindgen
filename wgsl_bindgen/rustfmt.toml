indent_style = "Block"
reorder_imports = true
tab_spaces = 2
max_width=90
fn_call_width=80
group_imports="StdExternalCrate"
