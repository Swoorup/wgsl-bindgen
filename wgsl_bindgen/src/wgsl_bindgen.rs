use std::io::Write;
use std::path::PathBuf;

use derive_builder::Builder;
use miette::Diagnostic;
use naga_oil::compose::{
  ComposableModuleDescriptor, Composer, ComposerError, NagaModuleDescriptor,
  ShaderLanguage,
};
use thiserror::Error;

use self::source_file::SourceFile;
use crate::{
  bevy_util::*, WgslEntryResult, WgslTypeMap, WgslTypeMapBuild, WgslTypeSerializeStrategy,
};
use crate::{create_rust_bindings, CreateModuleError, SourceFilePath};

const PKG_VER: &str = env!("CARGO_PKG_VERSION");
const PKG_NAME: &str = env!("CARGO_PKG_NAME");

/// Enum representing the possible errors that can occur in the `wgsl_bindgen` process.
///
/// This enum is used to represent all the different kinds of errors that can occur
/// when parsing WGSL shaders, generating Rust bindings, or performing other operations
/// in `wgsl_bindgen`.
#[derive(Debug, Error, Diagnostic)]
pub enum WgslBindgenError {
  #[error("All required fields need to be set upfront: {0}")]
  OptionBuilderError(#[from] WgslBindgenOptionBuilderError),

  #[error(transparent)]
  #[diagnostic(transparent)]
  DependencyTreeError(#[from] DependencyTreeError),

  #[error("Failed to compose modules with entry `{entry}`\n{msg}")]
  NagaModuleComposeError {
    entry: String,
    msg: String,
    inner: naga_oil::compose::ComposerErrorInner,
  },

  #[error(transparent)]
  ModuleCreationError(#[from] CreateModuleError),

  #[error(transparent)]
  WriteOutputError(#[from] std::io::Error),

  #[error("Output file is not specified. Maybe use `generate_string` instead")]
  OutputFileNotSpecified,
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
pub enum WgslShaderSourceOutputType {
  /// Include the final shader string directly in the output
  #[default]
  FinalShaderString,

  /// Use Composer including helper functions which will be executed on runtime
  Composer,
}

/// A struct representing a directory to scan for additional source files.
///
/// This struct is used to represent a directory to scan for additional source files
/// when generating Rust bindings for WGSL shaders. The `module_import_root` field
/// is used to specify the root prefix or namespace that should be applied to all
/// shaders given as the entrypoints, and the `directory` field is used to specify
/// the directory to scan for additional source files.
#[derive(Debug, Clone, Default)]
pub struct AdditionalScanDirectory {
  pub module_import_root: Option<String>,
  pub directory: String,
}

impl From<(Option<&str>, &str)> for AdditionalScanDirectory {
  fn from((module_import_root, directory): (Option<&str>, &str)) -> Self {
    Self {
      module_import_root: module_import_root.map(ToString::to_string),
      directory: directory.to_string(),
    }
  }
}

#[derive(Debug, Default, Builder)]
#[builder(
  setter(into),
  field(private),
  build_fn(private, name = "fallible_build")
)]
pub struct WgslBindgenOption {
  /// A vector of entry points to be added. Each entry point is represented as a `String`.
  #[builder(setter(each(name = "add_entry_point", into)))]
  pub entry_points: Vec<String>,

  /// The root prefix/namespace if any applied to all shaders given as the entrypoints.
  #[builder(default, setter(strip_option, into))]
  pub module_import_root: Option<String>,

  /// A boolean flag indicating whether to emit a rerun-if-changed directive to Cargo. Defaults to `true`.
  #[builder(default = "true")]
  pub emit_rerun_if_change: bool,

  /// A boolean flag indicating whether to skip header comments. Enabling headers allows to not rerun if contents did not change.
  #[builder(default = "false")]
  pub skip_header_comments: bool,

  /// A boolean flag indicating whether to skip the hash check. This will avoid reruns of bindings generation if
  /// entry shaders including their imports has not changed. Defaults to `false`.
  #[builder(default = "false")]
  pub skip_hash_check: bool,

  /// Derive [encase::ShaderType](https://docs.rs/encase/latest/encase/trait.ShaderType.html#)
  /// for user defined WGSL structs when `WgslTypeSerializeStrategy::Encase`.
  /// else derive bytemuck
  #[builder(default)]
  pub serialization_strategy: WgslTypeSerializeStrategy,

  /// Derive [serde::Serialize](https://docs.rs/serde/1.0.159/serde/trait.Serialize.html)
  /// and [serde::Deserialize](https://docs.rs/serde/1.0.159/serde/trait.Deserialize.html)
  /// for user defined WGSL structs when `true`.
  #[builder(default = "false")]
  pub derive_serde: bool,

  /// The type of output for the shader source. Defaults to `FinalShaderString`.
  #[builder(default)]
  pub shader_source_output_type: WgslShaderSourceOutputType,

  /// A mapping operation for WGSL built-in types. This is used to map WGSL built-in types to their corresponding representations.
  #[builder(setter(custom))]
  pub wgsl_type_map: WgslTypeMap,

  /// The output file path for the generated Rust bindings. Defaults to `None`.
  #[builder(default, setter(strip_option, into))]
  pub output_file: Option<PathBuf>,

  /// The additional set of directories to scan for source files.
  #[builder(default, setter(into, each(name = "additional_scan_dir", into)))]
  pub additional_scan_dirs: Vec<AdditionalScanDirectory>,
}

impl WgslBindgenOptionBuilder {
  pub fn build(&self) -> Result<WGSLBindgen, WgslBindgenError> {
    let options = self.fallible_build()?;
    WGSLBindgen::new(options)
  }

  pub fn wgsl_type_map(&mut self, map_build: impl WgslTypeMapBuild) -> &mut Self {
    let serialization_strategy = self
      .serialization_strategy
      .expect("Serialization strategy must be set before `wgs_type_map`");

    self.wgsl_type_map = Some(map_build.build(serialization_strategy));
    self
  }
}

pub struct WGSLBindgen {
  dependency_tree: DependencyTree,
  options: WgslBindgenOption,
  content_hash: String,
}

impl WGSLBindgen {
  fn new(options: WgslBindgenOption) -> Result<Self, WgslBindgenError> {
    let entry_points = options
      .entry_points
      .iter()
      .cloned()
      .map(SourceFilePath::new)
      .collect();

    let dependency_tree = DependencyTree::try_build(
      options.module_import_root.clone(),
      entry_points,
      options.additional_scan_dirs.clone(),
    )?;

    let content_hash = Self::get_contents_hash(&options, &dependency_tree);

    if options.emit_rerun_if_change {
      for file in Self::iter_files_to_watch(&dependency_tree) {
        println!("cargo:rerun-if-changed={}", file);
      }
    }

    Ok(Self {
      dependency_tree,
      options,
      content_hash,
    })
  }

  fn iter_files_to_watch(dep_tree: &DependencyTree) -> impl Iterator<Item = String> {
    dep_tree
      .all_files_including_dependencies()
      .into_iter()
      .map(|path| path.to_string())
  }

  fn get_contents_hash(options: &WgslBindgenOption, dep_tree: &DependencyTree) -> String {
    let mut hasher = blake3::Hasher::new();

    hasher.update(format!("{:?}", options).as_bytes());
    hasher.update(PKG_VER.as_bytes());

    for SourceFile { content, .. } in dep_tree.parsed_files() {
      hasher.update(content.as_bytes());
    }

    hasher.finalize().to_string()
  }

  fn generate_naga_module_for_entry(
    entry: SourceWithFullDependenciesResult<'_>,
  ) -> Result<WgslEntryResult, WgslBindgenError> {
    let map_err = |composer: &Composer, err: ComposerError| {
      let msg = err.emit_to_string(composer);
      WgslBindgenError::NagaModuleComposeError {
        entry: entry.source_file.file_path.to_string(),
        inner: err.inner,
        msg,
      }
    };

    let mut composer = Composer::default();
    let source = entry.source_file;

    for dependency in entry.full_dependencies.iter() {
      composer
        .add_composable_module(ComposableModuleDescriptor {
          source: &dependency.content,
          file_path: &dependency.file_path.to_string(),
          language: ShaderLanguage::Wgsl,
          as_name: dependency.module_name.as_ref().map(|name| name.to_string()),
          ..Default::default()
        })
        .map(|_| ())
        .map_err(|err| map_err(&composer, err))?;
    }

    let module = composer
      .make_naga_module(NagaModuleDescriptor {
        source: &source.content,
        file_path: &source.file_path.to_string(),
        ..Default::default()
      })
      .map_err(|err| map_err(&composer, err))?;

    Ok(WgslEntryResult {
      mod_name: source.file_path.file_prefix(),
      naga_module: module,
      source_including_deps: entry,
    })
  }

  pub fn generate_string(&self) -> Result<String, WgslBindgenError> {
    use std::fmt::Write;
    let entry_results = self
      .dependency_tree
      .get_source_files_with_full_dependencies()
      .into_iter()
      .map(Self::generate_naga_module_for_entry)
      .collect::<Result<Vec<_>, _>>()?;

    let mut text = String::new();

    if !self.options.skip_header_comments {
      writeln!(&mut text, "// File automatically generated by {PKG_NAME}^").unwrap();
      writeln!(&mut text, "//").unwrap();
      writeln!(&mut text, "// ^ {PKG_NAME} version {PKG_VER}",).unwrap();
      writeln!(&mut text, "// Changes made to this file will not be saved.").unwrap();
      writeln!(&mut text, "// SourceHash: {}", self.content_hash).unwrap();
      writeln!(&mut text).unwrap();
    }

    let output = create_rust_bindings(entry_results, &self.options)?;
    text += &output;

    Ok(text)
  }

  pub fn generate(&self) -> Result<(), WgslBindgenError> {
    let output_path = self
      .options
      .output_file
      .as_ref()
      .ok_or(WgslBindgenError::OutputFileNotSpecified)?;

    let old_content =
      std::fs::read_to_string(output_path).unwrap_or_else(|_| String::new());

    let old_hashstr_comment = old_content
      .lines()
      .find(|line| line.starts_with("// SourceHash:"))
      .unwrap_or("");

    let is_hash_changed =
      || old_hashstr_comment != format!("// SourceHash: {}", &self.content_hash);

    if self.options.skip_hash_check || is_hash_changed() {
      let content = self.generate_string()?;
      std::fs::File::create(output_path)?.write_all(content.as_bytes())?
    }

    Ok(())
  }
}
