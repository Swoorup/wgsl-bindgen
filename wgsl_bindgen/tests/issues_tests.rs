use std::fs::read_to_string;

use miette::{IntoDiagnostic, Result};
use pretty_assertions::assert_eq;
use wgsl_bindgen::*;

#[test]
fn test_issue_35() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("test/shaders/issue_35")
    .add_entry_point("tests/shaders/issue_35/clear.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .short_constructor(2)
    .shader_source_type(WgslShaderSourceType::UseComposerEmbed)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issue_35.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual = read_to_string("tests/output/issue_35.actual.rs").unwrap();
  let expected = read_to_string("tests/output/issue_35.expected.rs").unwrap();

  assert_eq!(actual, expected);
  Ok(())
}
