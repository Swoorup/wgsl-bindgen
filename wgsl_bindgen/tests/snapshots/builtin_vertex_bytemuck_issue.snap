---
source: wgsl_bindgen/tests/issues_tests.rs
---
#![allow(unused, non_snake_case, non_camel_case_types, non_upper_case_globals)]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ShaderEntry {
  BuiltinVertexBytemuck,
}
impl ShaderEntry {
  pub fn create_pipeline_layout(&self, device: &wgpu::Device) -> wgpu::PipelineLayout {
    match self {
      Self::BuiltinVertexBytemuck => {
        builtin_vertex_bytemuck::create_pipeline_layout(device)
      }
    }
  }
  pub fn create_shader_module_embed_source(
    &self,
    device: &wgpu::Device,
  ) -> wgpu::ShaderModule {
    match self {
      Self::BuiltinVertexBytemuck => {
        builtin_vertex_bytemuck::create_shader_module_embed_source(device)
      }
    }
  }
}
mod _root {
  pub use super::*;
}
pub mod layout_asserts {
  use super::{_root, _root::*};
  const WGSL_BASE_TYPE_ASSERTS: () = {
    assert!(std::mem::size_of::<glam::Vec3A>() == 16);
    assert!(std::mem::align_of::<glam::Vec3A>() == 16);
    assert!(std::mem::size_of::<glam::Vec4>() == 16);
    assert!(std::mem::align_of::<glam::Vec4>() == 16);
    assert!(std::mem::size_of::<glam::Mat3A>() == 48);
    assert!(std::mem::align_of::<glam::Mat3A>() == 16);
    assert!(std::mem::size_of::<glam::Mat4>() == 64);
    assert!(std::mem::align_of::<glam::Mat4>() == 16);
  };
}
pub mod builtin_vertex_bytemuck {
  use super::{_root, _root::*};
  #[repr(C)]
  #[derive(Debug, PartialEq, Clone, Copy)]
  pub struct VertexInput {}
  impl VertexInput {
    pub const fn new() -> Self {
      Self {}
    }
  }
  impl VertexInput {
    pub const VERTEX_ATTRIBUTES: [wgpu::VertexAttribute; 0] = [];
    pub const fn vertex_buffer_layout(
      step_mode: wgpu::VertexStepMode,
    ) -> wgpu::VertexBufferLayout<'static> {
      wgpu::VertexBufferLayout {
        array_stride: std::mem::size_of::<Self>() as u64,
        step_mode,
        attributes: &Self::VERTEX_ATTRIBUTES,
      }
    }
  }
  pub const ENTRY_VS_MAIN: &str = "vs_main";
  #[derive(Debug)]
  pub struct VertexEntry<const N: usize> {
    pub entry_point: &'static str,
    pub buffers: [wgpu::VertexBufferLayout<'static>; N],
    pub constants: Vec<(&'static str, f64)>,
  }
  pub fn vertex_state<'a, const N: usize>(
    module: &'a wgpu::ShaderModule,
    entry: &'a VertexEntry<N>,
  ) -> wgpu::VertexState<'a> {
    wgpu::VertexState {
      module,
      entry_point: Some(entry.entry_point),
      buffers: &entry.buffers,
      compilation_options: wgpu::PipelineCompilationOptions {
        constants: &entry.constants,
        ..Default::default()
      },
    }
  }
  pub fn vs_main_entry(vertex_input: wgpu::VertexStepMode) -> VertexEntry<1> {
    VertexEntry {
      entry_point: ENTRY_VS_MAIN,
      buffers: [VertexInput::vertex_buffer_layout(vertex_input)],
      constants: Default::default(),
    }
  }
  #[derive(Debug)]
  pub struct WgpuPipelineLayout;
  impl WgpuPipelineLayout {
    pub fn bind_group_layout_entries(
      entries: [wgpu::BindGroupLayout; 0],
    ) -> [wgpu::BindGroupLayout; 0] {
      entries
    }
  }
  pub fn create_pipeline_layout(device: &wgpu::Device) -> wgpu::PipelineLayout {
    device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
      label: Some("BuiltinVertexBytemuck::PipelineLayout"),
      bind_group_layouts: &[],
      push_constant_ranges: &[],
    })
  }
  pub fn create_shader_module_embed_source(device: &wgpu::Device) -> wgpu::ShaderModule {
    let source = std::borrow::Cow::Borrowed(SHADER_STRING);
    device.create_shader_module(wgpu::ShaderModuleDescriptor {
      label: Some("builtin_vertex_bytemuck.wgsl"),
      source: wgpu::ShaderSource::Wgsl(source),
    })
  }
  pub const SHADER_STRING: &str = r#"
struct VertexInput {
    @builtin(vertex_index) vertex_index: u32,
}

@vertex 
fn vs_main(input: VertexInput) -> @builtin(position) vec4<f32> {
    return vec4<f32>(0f, 0f, 0f, 1f);
}
"#;
}
pub mod bytemuck_impls {
  use super::{_root, _root::*};
  unsafe impl bytemuck::Zeroable for builtin_vertex_bytemuck::VertexInput {}
  unsafe impl bytemuck::Pod for builtin_vertex_bytemuck::VertexInput {}
}
